GH-41973: Expose new S3 option check_directory_existence_before_creation

HaochengLIU opened the pull request:
<!--
Thanks for opening a pull request!
If this is your first pull request you can find detailed information on how 
to contribute here:
  * [New Contributor's Guide](https://arrow.apache.org/docs/dev/developers/guide/step_by_step/pr_lifecycle.html#reviews-and-merge-of-the-pull-request)
  * [Contributing Overview](https://arrow.apache.org/docs/dev/developers/overview.html)


If this is not a [minor PR](https://github.com/apache/arrow/blob/main/CONTRIBUTING.md#Minor-Fixes). Could you open an issue for this pull request on GitHub? https://github.com/apache/arrow/issues/new/choose

Opening GitHub issues ahead of time contributes to the [Openness](http://theapacheway.com/open/#:~:text=Openness%20allows%20new%20users%20the,must%20happen%20in%20the%20open.) of the Apache Arrow project.

Then could you also rename the pull request title in the following format?

    GH-${GITHUB_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}

or

    MINOR: [${COMPONENT}] ${SUMMARY}

In the case of PARQUET issues on JIRA the title also supports:

    PARQUET-${JIRA_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}

-->

### Rationale for this change
Expose new S3 option `check_directory_existence_before_creation` from GH-41493
[PR](https://github.com/apache/arrow/pull/41972) to expose in Python is also under review.
<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

### What changes are included in this PR?
Expose new S3 option `check_directory_existence_before_creation` from GH-41493
<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

### Are these changes tested?
yes
<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

### Are there any user-facing changes?
Yes. R function documentation is updated.
<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please uncomment the line below and explain which changes are breaking.
-->
<!-- **This PR includes breaking changes to public APIs.** -->

<!--
Please uncomment the line below (and provide explanation) if the changes fix either (a) a security vulnerability, (b) a bug that caused incorrect or invalid data to be produced, or (c) a bug that causes a crash (even when the API contract is upheld). We use this to highlight fixes to issues that may affect users without their knowledge. For this reason, fixing bugs that cause errors don't count, since those are usually obvious.
-->
<!-- **This PR contains a "Critical Fix".** -->
* GitHub Issue: #41973

HaochengLIU commented:
Hi @nealrichardson , could you pls review when you have time? I have to admit I use C++/Python in my daily life but know nothing about R and just follow the existing code for allow_bucket_deletion to expose it.

HaochengLIU commented:
Hi @paleolimbot @thisisnic , could you review this PR when you have time? Thanks.

amoeba commented:
Hi @HaochengLIU, I took a look through and left some suggestions.

One thing I think we'll have to sort out here is the issue that's making the [Check minimum supported Arrow C++ Version](https://github.com/apache/arrow/actions/runs/9393455224/job/25869530975?pr=41998) check fail. We'll need an ifdef guard in `r/src/filesystem.cpp` but I'm not sure if we've got a pattern for what the user experiences.

HaochengLIU commented:
Hi @amoeba , I rebased and the somehow do not see the suggestion you are referring to.. Could you point me to that?
Also could you guide what do you mean by "ifdef guard" in the filesystem.cpp? It's not using any standard or platform dependent code, simply a new argument 

Per the install error 
```
* checking whether package ‘arrow’ can be installed ... [87s/90s] ERROR
Installation failed.
See ‘/home/runner/work/arrow/arrow/src/r/check/arrow.Rcheck/00install.out’ for details.
```
Where is this 00install.out log 😢 
ty.

thisisnic commented:
@HaochengLIU The CI output there can be a bit tricky to follow, but the important bit is in the next section: https://github.com/apache/arrow/actions/runs/10207786938/job/28243188693?pr=41998#step:8:49

amoeba commented:
Sorry @HaochengLIU, I think I forgot to save my review. You should see them now.

@thisisnic is right about my note. See this example: https://github.com/apache/arrow/blob/c69c1d80eb27022c1b3f91fbbc7979f8293c9e50/r/src/extension-impl.cpp#L103

I think here we want to do here is (1) conditionally define two versions of `fs___S3FileSystem__create`, one that supports `check_directory_existence_before_creation` and one that doesn't and then (2) catch calls to `S3FileSystem$create` which pass the option `check_directory_existence_before_creation` when `arrow::arrow_info()$build_info$cpp_version < "17.0.0"` and probably emit a warning but otherwise continue. Let me know if this isn't clear.

nealrichardson commented:
> Sorry @HaochengLIU, I think I forgot to save my review. You should see them now.
> 
> @thisisnic is right about my note. See this example:
> 
> https://github.com/apache/arrow/blob/c69c1d80eb27022c1b3f91fbbc7979f8293c9e50/r/src/extension-impl.cpp#L103
> 
> I think here we want to do here is (1) conditionally define two versions of `fs___S3FileSystem__create`, one that supports `check_directory_existence_before_creation` and one that doesn't and then (2) catch calls to `S3FileSystem$create` which pass the option `check_directory_existence_before_creation` when `arrow::arrow_info()$build_info$cpp_version < "17.0.0"` and probably emit a warning but otherwise continue. Let me know if this isn't clear.

The tricky thing about that is the codegen script that creates `arrowExports.cpp`: it doesn't know how to handle the conditional definitions like that, and I'm not sure it's worth extending to do so.

Another option would be to refactor to send a list of options to `fs___S3FileSystem__create` (and the other FileSystem create methods, for consistency) and then pull the settings out of the list inside the C++ function, like how `make_compute_options()` works (https://github.com/apache/arrow/blob/main/r/src/compute.cpp#L129). Then the `#if ARROW_VERSION_MAJOR >= 16` would just be wrapping a few lines inside that function and not affecting the signature.

HaochengLIU commented:
Hi folks, thanks for reviewing and I've addressed most of the comments. 
Referring to the non-backward compatible `check_directory_existence_before_creation` flag, IMHO refactoring fs___S3FileSystem__create (and the other FileSystem create methods, for consistency) sounds should be a seperate PR

Though if [arrow/r/src/extension-impl.cpp](https://github.com/apache/arrow/blob/c69c1d80eb27022c1b3f91fbbc7979f8293c9e50/r/src/extension-impl.cpp#L103) will not work per @nealrichardson , I can try to tackle the refactor first once I'm back from vacation

amoeba commented:
Thanks for the catch @nealrichardson and the suggestion. I'd be supportive of approving this PR with just the "Check minimum supported Arrow C++ Version " job failing and just my existing suggestions applied.

HaochengLIU commented:
Hi, sorry for the delay here, was carried away by other life events. Let me know if there is anything i can help here

thisisnic commented:
> Hi, sorry for the delay here, was carried away by other life events. Let me know if there is anything i can help here

Not your fault at all, we didn't resolve the earlier discussion on previous versions.  Based on the CI results after rebasing the PR, there are a few changes needed before merging this, which I was going to take a look at myself.  If you'd rather take a look though, let me know, and I'll leave it to you instead!

HaochengLIU commented:
> > Hi, sorry for the delay here, was carried away by other life events. Let me know if there is anything i can help here
> 
> Not your fault at all, we didn't resolve the earlier discussion on previous versions. Based on the CI results after rebasing the PR, there are a few changes needed before merging this, which I was going to take a look at myself. If you'd rather take a look though, let me know, and I'll leave it to you instead!

I will be in a national park without wifi for the incoming days, could I handle this PR to you :) 
Much appreciated!
